File: csharp/ToolGood.Algorithm/UnitConversion/AreaConverter.cs
using System;
using System.Collections.Generic;

namespace ToolGood.Algorithm.UnitConversion
{
    internal sealed class AreaConverter 
    {
		private static readonly Dictionary<string, decimal> units2 = new Dictionary<string, decimal>(StringComparer.OrdinalIgnoreCase) {
			{"m²",1 }, {"m2",1 }, {"square metre",1 },{"square meter",1 },{"centiare",1 },{"平方米",1 },{"平方公尺",1 },
			{"km²",0.000001m }, {"km2",0.000001m }, {"square kilometre",0.000001m },{"square kilometer",0.000001m },{"平方千米",0.000001m },
			{"dm²",100 }, {"dm2",100 }, {"square decimetre",100 },{"square decimeter",100 },{"平方分米",100 },
			{"cm²",10000 }, {"cm2",10000 }, {"square centimetre",10000 },{"square centimeter",10000 },{"平方厘米",10000 },
			{"mm²",1000000 }, {"mm2",1000000 }, {"square millimetre",1000000 },{"square millimeter",1000000 },{"平方毫米",1000000 },
			{"ft²",1m /  0.3048m /  0.3048m }, {"ft2",1m /  0.3048m /  0.3048m }, {"square foot",1m /  0.3048m /  0.3048m },{"square feet",1m /  0.3048m /  0.3048m },{"sq ft",1m /  0.3048m /  0.3048m },{"平方英尺",1m /  0.3048m /  0.3048m },
			{"yd²",1m /  0.9144m /  0.9144m }, {"yd2",1m /  0.9144m /  0.9144m }, {"sq yd",1m /  0.9144m /  0.9144m },{"square yard",1m /  0.9144m /  0.9144m },{"平方码",1m /  0.9144m /  0.9144m },
			{"a",0.01m  },{"are",0.01m  },
			{"ha",0.0001m }, {"hectare",0.0001m }, {"公顷",0.0001m },
			{"in²",1m / 0.00064516m }, {"in2",1m / 0.00064516m }, {"sq in",1m / 0.00064516m },{"square inch",1m / 0.00064516m },{"平方英寸",1m / 0.00064516m },
			{"mi²",1m / 2589988.110336m }, {"mi2",1m / 2589988.110336m }, {"sq mi",1m / 2589988.110336m },{"square mile",1m / 2589988.110336m },{"平方英里",1m / 2589988.110336m },
			{"亩",1m / 666.667m  },
		};

		public static bool TryConvert(string leftSynonym, string rightSynonym, decimal left, out decimal right)
		{
			if(units2.TryGetValue(leftSynonym, out decimal l)) {
				if(units2.TryGetValue(rightSynonym, out decimal r)) {
					right = left / l * r;
					return true;
				}
			}
			right = decimal.Zero;
			return false;
		}

	 
    }
}

File: java/toolgood.algorithm/src/main/java/toolgood/algorithm/unitConversion/AreaConverter.java
package toolgood.algorithm.unitConversion;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.Map;
import java.util.TreeMap;

public final class AreaConverter {
    private static final Map<String, BigDecimal> units2;

    static {
        TreeMap<String, BigDecimal> map = new TreeMap<>(String.CASE_INSENSITIVE_ORDER);
        map.put("m²", BigDecimal.ONE);
        map.put("m2", BigDecimal.ONE);
        map.put("square metre", BigDecimal.ONE);
        map.put("square meter", BigDecimal.ONE);
        map.put("centiare", BigDecimal.ONE);
        map.put("平方米", BigDecimal.ONE);
        map.put("平方公尺", BigDecimal.ONE);

        map.put("km²", new BigDecimal("0.000001"));
        map.put("km2", new BigDecimal("0.000001"));
        map.put("square kilometre", new BigDecimal("0.000001"));
        map.put("square kilometer", new BigDecimal("0.000001"));
        map.put("平方千米", new BigDecimal("0.000001"));

        map.put("dm²", new BigDecimal("100"));
        map.put("dm2", new BigDecimal("100"));
        map.put("square decimetre", new BigDecimal("100"));
        map.put("square decimeter", new BigDecimal("100"));
        map.put("平方分米", new BigDecimal("100"));

        map.put("cm²", new BigDecimal("10000"));
        map.put("cm2", new BigDecimal("10000"));
        map.put("square centimetre", new BigDecimal("10000"));
        map.put("square centimeter", new BigDecimal("10000"));
        map.put("平方厘米", new BigDecimal("10000"));

        map.put("mm²", new BigDecimal("1000000"));
        map.put("mm2", new BigDecimal("1000000"));
        map.put("square millimetre", new BigDecimal("1000000"));
        map.put("square millimeter", new BigDecimal("1000000"));
        map.put("平方毫米", new BigDecimal("1000000"));

        BigDecimal sqft = BigDecimal.ONE.divide(new BigDecimal("0.3048"), 30, RoundingMode.HALF_UP)
                .divide(new BigDecimal("0.3048"), 30, RoundingMode.HALF_UP);
        map.put("ft²", sqft);
        map.put("ft2", sqft);
        map.put("square foot", sqft);
        map.put("square feet", sqft);
        map.put("sq ft", sqft);
        map.put("平方英尺", sqft);

        BigDecimal sqyd = BigDecimal.ONE.divide(new BigDecimal("0.9144"), 30, RoundingMode.HALF_UP)
                .divide(new BigDecimal("0.9144"), 30, RoundingMode.HALF_UP);
        map.put("yd²", sqyd);
        map.put("yd2", sqyd);
        map.put("sq yd", sqyd);
        map.put("square yard", sqyd);
        map.put("平方码", sqyd);

        map.put("a", new BigDecimal("0.01"));
        map.put("are", new BigDecimal("0.01"));

        map.put("ha", new BigDecimal("0.0001"));
        map.put("hectare", new BigDecimal("0.0001"));
        map.put("公顷", new BigDecimal("0.0001"));

        BigDecimal sqin = BigDecimal.ONE.divide(new BigDecimal("0.00064516"), 30, RoundingMode.HALF_UP);
        map.put("in²", sqin);
        map.put("in2", sqin);
        map.put("sq in", sqin);
        map.put("square inch", sqin);
        map.put("平方英寸", sqin);

        BigDecimal sqmi = BigDecimal.ONE.divide(new BigDecimal("2589988.110336"), 30, RoundingMode.HALF_UP);
        map.put("mi²", sqmi);
        map.put("mi2", sqmi);
        map.put("sq mi", sqmi);
        map.put("square mile", sqmi);
        map.put("平方英里", sqmi);

        BigDecimal mu = BigDecimal.ONE.divide(new BigDecimal("666.667"), 30, RoundingMode.HALF_UP);
        map.put("亩", mu);

        units2 = map;
    }

    public static BigDecimal TryConvert(String leftSynonym, String rightSynonym, BigDecimal left) {
        BigDecimal l = units2.get(leftSynonym);
        if (l != null) {
            BigDecimal r = units2.get(rightSynonym);
            if (r != null) {
                return left.divide(l, 30, RoundingMode.HALF_UP).multiply(r);
            }
        }
        return null;
    }
}
